---
title: "Annotation of clusters in 10x data"
author: "Arun Chavan"
output:
  html_notebook:
    toc: yes
    toc_float: yes
    number_sections: true
bibliography: ../refs.bib
---
Started: 2020-09-02  
Last edited: `r format(Sys.time())`

```{r message=FALSE, warning=FALSE}
### packages ==================================================================
library(tidyverse)

# single cell
library(Seurat)

# rmd
library(kableExtra)

# plotting
library(patchwork)
library(ggthemes)

# go enrichment
library(clusterProfiler)
library(org.Hs.eg.db)
```

We can take a first pass at annotating the clusters from our 10x data by comparing them to the reference dataset that we put together. This was done by compiling celltype-averaged expression values from the following single cell RNA-seq studies: [@pavlicev_single-cell_2017; @vento-tormo_single-cell_2018; @suryawanshi_single-cell_2018]. 

# Data

## Samples
This is the sample information sent by Alice. Not sure about the INP id for `AL09` and `AL10`. The library for `INP188` villi (associated with `AL07`) was problematic, so is not used.  

```{r}
sample.info <- read.delim("../info/from-alice/sample_info.tsv", stringsAsFactors = FALSE)
sample.info %>% kable(caption = "Sample information") %>% kable_styling(full_width = FALSE)
```

## Seurat-processed data
Read the `Seurat`-processed (by Eric) data sent by Alice.  

```{r}
seur <- readRDS("../info/from-alice/placenta.rds")

head(seur@meta.data)
```
### Add sample metadata

The `orig.ident` column in the metadata represents the sample IDs from the sample info table above. We just have rename them to left-pad the numbers for consistency, and set new cluster names as default idents.

```{r}
# left pad orig.idents
seur@meta.data$orig.ident <- gsub("(AL)(\\d)$", "\\10\\2", seur@meta.data$orig.ident)

# add additional metadata (covid status, tissue)
seur@meta.data$covid <- sample.info$covid[match(x = seur@meta.data$orig.ident, 
                                                table = sample.info$sample_id)]
seur@meta.data$tissue <- sample.info$tissue[match(x = seur@meta.data$orig.ident,
                                                  table = sample.info$sample_id)]

# rename clusters for consistency
seur$seurat_clusters <- paste0("clust_", seur$seurat_clusters)
seur$seurat_clusters <- gsub("(clust_)(\\d$)", "\\10\\2", as.character(seur$seurat_clusters)) # left pad
seur$seurat_clusters <- factor(seur$seurat_clusters, levels = paste0("clust_", c("00", "01", "02", "03", "04", "05", "06", "07", "08", "09", 10:34)))

# set seurat clusters as default idents
Idents(seur) <- seur@meta.data$seurat_clusters
```

```{r}
# add another column to metadata with renames orig.idents as COVID_1, CNTRL_1 instead of AL01, AL02, etc.
AL_covid_mapping <- unique(seur@meta.data[, c("orig.ident", "covid")])

AL_covid_mapping$renamed <- NA
counter.covid <- 1
counter.cntrl <- 1
for(i in 1:nrow(AL_covid_mapping)) {
  
  if(AL_covid_mapping$covid[i] == "covid") {
    AL_covid_mapping$renamed[i] <- paste0("COVID_", counter.covid)
    counter.covid <- counter.covid + 1 
  }
  
  if(AL_covid_mapping$covid[i] == "cntrl") {
    AL_covid_mapping$renamed[i] <- paste0("CNTRL_", counter.cntrl)
    counter.cntrl <- counter.cntrl + 1
  }
}

seur@meta.data$orig.ident_renamed <- AL_covid_mapping$renamed[
  match(x = seur@meta.data$orig.ident, 
        table = AL_covid_mapping$orig.ident)
]
```

## Average by cluster
We have to get transcriptomes averaged by clusters, so that we can compare them with the reference data in order to narrow down annotations. For averaging, we will use `sctransform` normalized data since this is what we use for marker gene detection down the line.

```{r message=FALSE}
# average sctransformed expression
seur.avg <- AverageExpression(seur)

# averaged data to use for correlations (SCT transformed)
plac <- seur.avg$SCT
```

## Reference data
```{r}
# read reference datasets
ref <- read.csv("../results/01_reference-atlas/vento-surya-pavli_joined.csv", stringsAsFactors = FALSE)
```

## Get data in shape
We will join the data with the reference data so they are both in the same dataframe. 

```{r}
# add gene names column
plac$gene_name <- rownames(plac)

# inner join
plac <- dplyr::inner_join(plac, ref, by = "gene_name")

head(plac)
```

# Correlations within the data
Before moving to annotating clusters by comparison with the reference dataset, first we will look at the correlation structure within the dataset.

## Correlation matrix
```{r fig.asp=0.7}
# build correlation matrix from expression data
cor.matrix <- cor(plac %>% dplyr::select(starts_with("clust")), method = 'spearman')

# reorder correlation matrix based on clustering
dd <- as.dist((1 - cor.matrix)) 
hc <- hclust(dd, method = 'complete')
cormat <- cor.matrix[hc$order, hc$order]

# melt correlation matrix
dat <- reshape2::melt(cormat, na.rm = T)

## plot
p <- ggplot(data = dat, aes(Var1, Var2, fill = value)) +
  geom_tile(colour = "white") +
  scale_fill_gradient(low = 'white', high = 'red',
                      name = "Spearman\nCorrelation") +
  coord_fixed(ratio = 1) +
  labs(title = "Correlations among clusters") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, size = 8, 
                                   hjust=1, vjust = 0.5),
        axis.text.y = element_text(size=8),
        axis.ticks.length = unit(0.15, units = c('lines')),
        legend.title = element_text(size = 10),
        axis.title = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        legend.key.size = unit(0.8, units = c('lines')))

ggsave(p, filename = "../results/02_annotation/plots/corrplot_clusters.pdf", 
       device = "pdf", width = 7, height = 6, units = "in")

print(p)
```

## Heirarchical clustering
```{r}
# build correlation matrix from expression data
cor.matrix <- cor(plac %>% dplyr::select(starts_with("clust")), method = 'spearman')

# reorder correlation matrix based on clustering
dd <- as.dist((1 - cor.matrix)) 
hc <- hclust(dd, method = 'complete')

pdf(file = "../results/02_annotation/plots/hclust_clusters.pdf", width = 7, height = 5)
plot(hc, cex = 0.8)
dev.off()
plot(hc, cex = 0.8)
```

There are three big cluster blocks, with substructure within them. The cluster blocks also correspond well with the UMAP plot that was sent by Alice. 

# Annotation by correlation
Let's now actually measure correlations between all clusters with the reference datasets. We will treat our data as query dataset and for each cluster assign top 3 cell types in each reference dataset.

```{r}
refdata <- c("vento", "surya", "pavli")

# build correlation matrix from expression data
cor.matrix <- cor(plac[, names(plac)[names(plac) != "gene_name"]], 
                  method = 'spearman')

# reorder correlation matrix based on clustering
dd <- as.dist((1 - cor.matrix))
hc <- hclust(dd, method = 'complete')
cor.matrix <- cor.matrix[hc$order, hc$order]

# melt correlation matrix
cormat <- reshape2::melt(cor.matrix, na.rm = T)
cormat$Var1_source <- sapply(strsplit(as.character(cormat$Var1), split = "_"), "[[", 1)
cormat$Var2_source <- sapply(strsplit(as.character(cormat$Var2), split = "_"), "[[", 1)

# subset
cormat <- cormat[which(cormat$Var1_source %in% refdata & 
                   cormat$Var2_source == "clust"), ]

# top 3 matches with highest correlation coefficients
topmatch <- data.frame( # empty df to fill top matches from the loop below
  cluster = unique(as.character(cormat$Var2)) %>% sort(),
  vento.top1 = NA,
  surya.top1 = NA,
  pavli.top1 = NA,
  vento.top2 = NA,
  surya.top2 = NA,
  pavli.top2 = NA,
  vento.top3 = NA,
  surya.top3 = NA,
  pavli.top3 = NA
)

cormat$top3 <- NA

for(i in unique(cormat$Var2)){
  for(j in refdata){
    # identify top3 match indices
    ind <- which(cormat$Var2 == i & cormat$Var1_source == j)
    val <- cormat$value[ind]
    top3ind <- ind[order(val, decreasing = TRUE)[1:3]]
    
    # assign match ranking to correlation data for plotting
    cormat$top3[top3ind[1]] <- "1"
    cormat$top3[top3ind[2]] <- "2"
    cormat$top3[top3ind[3]] <- "3"
    
    # assign top matches to topmatches dataframe
    topmatch[topmatch$cluster == i, paste0(j, ".top1")] <- gsub(paste0(j, "_"), "", as.character(cormat$Var1[top3ind[1]]))
    topmatch[topmatch$cluster == i, paste0(j, ".top2")] <- gsub(paste0(j, "_"), "", as.character(cormat$Var1[top3ind[2]]))
    topmatch[topmatch$cluster == i, paste0(j, ".top3")] <- gsub(paste0(j, "_"), "", as.character(cormat$Var1[top3ind[3]]))
  }
}

```

## Plots

```{r}
# plotting function
clustAnnoPlot <- function(dat, query, reference){
  
  dat <- dat[which(dat$Var2_source == query & dat$Var1_source == reference), ]
  
  p <- ggplot(data = dat, 
              aes(Var1, Var2, fill = value)) +
    geom_tile(colour = "white") +
    scale_fill_gradient(low = 'white', high = 'red',
                        name = "Spearman\nCorrelation") +
    geom_point(aes(Var1, Var2, alpha = top3),
               size = 1.5, shape = 19, stroke  = 0) +
    scale_alpha_manual(values = c(1, 0.5, 0.25), 
                       breaks = c(1, 2, 3),
                       name = "top3", na.value = 0) +
    coord_fixed(ratio = 1) +
    xlab("reference") +
    ylab("query") +
    labs(caption = "For each celltype in query, black points represent top 3 celltypes from reference with highest correlation.",
         title = paste0(query, " vs. ", reference)) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, size = 8, 
                                     hjust=1, vjust = 0.5),
          axis.text.y = element_text(size=8),
          axis.ticks.length = unit(0.15, units = c('lines')),
          legend.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          plot.caption = element_text(size = 7),
          panel.grid = element_blank(),
          panel.border = element_blank(),
          legend.key.size = unit(0.8, units = c('lines')))
  
  return(p)
}

```

### Against Vento-Tormo et al
```{r fig.asp=0.9}
## Annotation plots
anno.vento <- clustAnnoPlot(dat = cormat, query = "clust", reference = "vento")
cowplot::ggsave2(anno.vento, device = "pdf", width = 7.5, height = 6.5, units = "in",
                 filename = "../results/02_annotation/plots/annotation-by-correlation_ref-vento.pdf")
print(anno.vento)
```

### Against Suryavanshi et al
```{r fig.asp=0.9}
## Annotation plots
anno.surya <- clustAnnoPlot(dat = cormat, query = "clust", reference = "surya")
cowplot::ggsave2(anno.surya, device = "pdf", width = 7.5, height = 6.5, units = "in",
                 filename = "../results/02_annotation/plots/annotation-by-correlation_ref-surya.pdf")
print(anno.surya)
```

### Against Pavlicev et al
```{r fig.asp=0.9}
## Annotation plots
anno.pavli <- clustAnnoPlot(dat = cormat, query = "clust", reference = "pavli")
cowplot::ggsave2(anno.pavli, device = "pdf", width = 7, height = 6, units = "in",
                 filename = "../results/02_annotation/plots/annotation-by-correlation_ref-pavli.pdf")
print(anno.pavli)
```

## Top matches
```{r}
topmatch %>% kable(caption = "Top 3 matches against all reference datasets.") %>% kable_styling(full_width = FALSE, font_size = 10)

```

## Labels for clusters
We need to create consistent labels for all cell types that we identify. Below is a `list` I created in which the top level objects are labels we will give to the cell types, within which contained are lists of corresponding labels from `vento` and `surya`. These are tentative labels. After the first pass, when we go through the clusters with a fine-toothed comb, we can modify them further. For example, if we have multiple clusters labeled as `dec.DSC` (decidual macrophages), and if those clusters are distinct, we can then break up this label into `dec.DSC1` and `dec.DSC2`. 

```{r}
labs <- list("dec.DSC"    = list("vento.lab" = c("dS1", "dS2", "dS3"),
                                 "surya.lab" = c("dec.DSC", "dec.FB1", "dec.FB2")),
             "DC"     = list("vento.lab" = c("DC1", "DC2"),
                                 "surya.lab" = c("dec.DC1", "dec.DC2")),
             "Mac"    = list("vento.lab" = c("dM1", "dM2", "dM3"),
                                 "surya.lab" = c("dec.MAC")),
             "NK"     = list("vento.lab" = c("dNK.p", "dNK1", "dNK2", "dNK3", "NK.CD16neg", "NK.CD16pos"),
                                 "surya.lab" = c("dec.NK1", "dec.NK2")),
             "dec.SMC"    = list("vento.lab" = c("dP1", "dP2"),
                                 "surya.lab" = c("dec.SMC")),
             "dec.Endo"   = list("vento.lab" = c("Endo.m"),
                                 "surya.lab" = c("dec.VEC")),
             "dec.Epi"    = list("vento.lab" = c("Epi1", "Epi2"),
                                 "surya.lab" = c("dec.EEC")),
             "Tcell"  = list("vento.lab" = c("Tcells"),
                                 "surya.lab" = c("dec.TC")),
             "vil.Endo"   = list("vento.lab" = c("Endo.f"),
                                 "surya.lab" = c("vil.VEC")),
             "vil.EVT"    = list("vento.lab" = c("EVT"),
                                 "surya.lab" = c("vil.EVT")),
             "vil.FB"     = list("vento.lab" = c("fFB1", "fFB2"),
                                 "surya.lab" = c("vil.FB1", "vil.FB2", "vil.FB3")),
             "vil.Hofb"   = list("vento.lab" = c("HB"),
                                 "surya.lab" = c("vil.HC")),
             "vil.SCT"    = list("vento.lab" = c("SCT"),
                                 "surya.lab" = c("vil.SCT")),
             "vil.VCT"    = list("vento.lab" = c("VCT"),
                                 "surya.lab" = c("vil.VCT")),
             "Gran"   = list("vento.lab" = c("Granulocytes"),
                                 "surya.lab" = c()),
             "ILC"    = list("vento.lab" = c("ILC3"),
                                 "surya.lab" = c()),
             "Mono"   = list("vento.lab" = c("MO"),
                                 "surya.lab" = c()),
             "Plasma" = list("vento.lab" = c("Plasma"),
                                 "surya.lab" = c()),
             "EB"     = list("vento.lab" = c(),
                                 "surya.lab" = c("vil.EB")),
             "unk.Endo.L" = list("vento.lab" = c("Endo.L"),
                                 "surya.lab" = c("dec.LEC"))
)

```

## Matches consistent between reference datasets
The following clusters have consistent top1 match between `vento` and `surya` references, i.e. they correspond to the same cell type category. The `pavli` reference is too unresolved to be used diagnostically, so we will ignore it for now. For some cell types it is confirmatory, though, e.g. cluster_00 corresponds to DSC in all three references. 

```{r}
# find out which clusters are consistent.
# If vento.top1 and surya.top1 are found in the same item in the labs list, the mapping is consistent.
consistent <- c(NA)
for(i in 1:nrow(topmatch)){
  consistent[i] <- grep(topmatch$vento.top1[i], labs) == grep(topmatch$surya.top1[i], labs)
}

# subset to include consistent set
topmatch.cons <- topmatch %>% 
  filter(consistent) %>% 
  dplyr::select(cluster, vento.top1, surya.top1)

# add our tentative labels to the clusters
for(i in 1:nrow(topmatch.cons)){
  topmatch.cons$label[i] <- names(labs)[grep(topmatch.cons$vento.top1[i], labs)]
}

# print
topmatch.cons[order(topmatch.cons$label), ] %>% knitr::kable() %>% kable_styling(full_width = FALSE)
```

## Ambiguous clusters
The top1 matches for some clusters with `vento` and `surya` are not the same. The are the clusters that will require a more detailed look to determine their identify. 

```{r}
# subset for ambiguous clusters
topmatch.ambig <- topmatch %>% 
  filter(!consistent) %>% 
  dplyr::select(cluster, vento.top1, surya.top1)

# add our labels.
for(i in 1:nrow(topmatch.ambig)){
  topmatch.ambig$label[i] <- paste0(
    names(labs)[grep(topmatch.ambig$vento.top1[i], labs)],
    " or ",
    names(labs)[grep(topmatch.ambig$surya.top1[i], labs)]
    )
}

# print
topmatch.ambig[order(topmatch.ambig$label), ] %>% kable() %>% kable_styling(full_width = FALSE)
```

# Annotation refinement
The annotations we have so far are only a starting point. Now we can look at each annotation more carefully to make sure that everything makes sense.

## Sample contribution to clusters
One of the ways in which we can refine the clusters further is by knowing which tissue that sample arises from. For example, if a cluster has macrophage gene signature and if most cells in that cluster come from villi samples, we can further narrow down the identity of the cluster to Hofbauer cells. We can do this by simply counting for each cluster how many cells come from decidua vs villi and by calculating the percentage. 

```{r fig.asp=1.2, fig.width=6.5}
# count cells in each cluster by tissue of origin
bytissue <- table(seur$tissue, seur$seurat_clusters) %>% 
  as.data.frame() %>% 
  dplyr::rename(tissue = Var1, cluster = Var2, frequency = Freq)

# calculate fraction
for(i in 1:nrow(bytissue)){
  bytissue$fraction[i] <- bytissue$frequency[i]/
    sum(bytissue$frequency[bytissue$cluster == bytissue$cluster[i]])
}

# plot
p.cells <- ggplot(data = bytissue, aes(x = frequency, y = cluster)) +
  geom_bar(aes(fill = tissue), stat = "identity", position = "stack") +
  ggtitle("Number of cells")

p.frac <- ggplot(data = bytissue, aes(x = fraction, y = cluster)) +
  geom_bar(aes(fill = tissue), stat = "identity", position = "stack") +
  ggtitle("Fraction of cells")

p.cells + p.frac + plot_layout(guides = "collect", nrow = 2) +
  plot_annotation(caption = "Sample contribution to clusters") &
  coord_flip() &
  scale_fill_tableau(palette = "Classic 10 Medium") &
  theme_minimal() +
  theme(panel.grid.minor = element_blank(),
        panel.grid.major = element_line(size = 0.25),
        plot.title = element_text(size = 10),
        axis.text.x = element_text(angle = 90, vjust = 0.5))
  

```
Following clusters have more than 70% cells coming from the same tissue of origin. 

```{r}
bytissue[, c("cluster", "tissue", "fraction")] %>% 
  pivot_wider(names_from = "tissue", values_from = "fraction") %>% 
  filter(decidua > 0.70 | villi > 0.70) %>% 
  kable(digits = 2, row.names = FALSE, 
        caption = "Clusters with more than 70% cells from one tissue") %>%
  kable_styling(full_width = FALSE)
```

Following clusters have ambiguous origin, i.e. they contain cells from decidua and villi samples. 

```{r}
bytissue[, c("cluster", "tissue", "fraction")] %>% 
  pivot_wider(names_from = "tissue", values_from = "fraction") %>% 
  filter(decidua < 0.70 & villi < 0.70) %>% 
  kable(digits = 2, row.names = FALSE, 
        caption = "Clusters with cells from both tissues") %>%
  kable_styling(full_width = FALSE)
```

This is largely consistent with the putative annotations. All putatively trophoblast clusters come overwhelmingly from villi samples. The only exception is extravillous trophoblast cluster, which as expected, comes from decidua samples. At least in the species that I have worked with, it is often difficult to separate the fetal tissue from maternal tissue. Therefore, we use this analysis only as a rough guide, i.e. when there is a conflict between the two, the marker gene-based annotation should take precedence over tissue of origin in cell type id because the latter is experimentally (and as a consequence, analytically) messy to disentangle.

Most clusters that have mixed origin, are putative immune cell type clusters. This could suggest that the immune cell types are present on both maternal and fetal sides of the interface, and because they are the same cell type, they end up in the same cluster. This means that we can't confidently say for immune cells whether they are decidual or villous. For this reason, when such ambiguity is present, we will label them without the tissue prefix, i.e. for example `Tcell` instead of `dec.Tcell` or `vil.Tcell`.

## Marker genes
### Identify marker genes 
I ran `Seurat::findAllMarkers` on this data using the `SCT` assay. This function outputs a list of marker genes for each cluster compared to all other cells in the data. This takes a while (hours) to run, so I ran it on the cluster and saved the results to a `.csv`. See the `code` directory for full code, and see below for the function call. 

```{r echo=TRUE, eval=FALSE}
# set default assay to SCT
DefaultAssay(object = plac) <- "SCT"

# find markers for all clusters
markers <- FindAllMarkers(object = plac, 
                          only.pos = TRUE,
                          logfc.threshold = 0.25)  

# write.csv
write.csv(markers, "/home/arc78/scratch60/covid-placenta/markers_sct.csv", row.names = FALSE)
```

```{r}
markers <- read.csv("../results/02_annotation/files/markers_sct.csv")

# rename clusters
markers$cluster <- paste0("clust_", markers$cluster)
markers$cluster <- gsub("(clust_)(\\d)$", "\\10\\2", markers$cluster)
markers$cluster <- factor(markers$cluster, levels = unique(markers$cluster))

# split by cluster
markers <- split.data.frame(markers, markers$cluster)
```

### Plot marker genes
We will plot top 50 marker genes for all clusters and save the plot to pdf. 
```{r}
# plotting function
DotPlot2 <- function(object = seur, assay = "SCT", features, title, ...) {
  p <- DotPlot(object, 
               assay = assay,
               features = features, 
               dot.min = 0.05, dot.scale = 4) +
    coord_flip() +
    labs(caption = paste0(assay, " normalized expression"), title = title) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
          panel.grid = element_blank())
  
  return(p)
}
```

```{r fig.asp=1}
top50marker.plots <- list()
for(i in names(markers)) {
  if(nrow(markers[[i]]) > 50)
    features <- markers[[i]]$gene[1:50]
  if(nrow(markers[[i]]) < 50)
    features <- markers[[i]]$gene

  top50marker.plots[[i]] <- DotPlot2(features = features, title = i)
         
  cowplot::ggsave2(top50marker.plots[[i]], 
                   filename = paste0("../results/02_annotation/plots/top50markers_dotplot_", i, ".pdf"),
                   width = 7.5, height = 7.5, units = "in")
}
```

# Manual annotation
With the marker gene plots we have made, we can examine the top marker genes in each cluster to manually confirm or adjust the annotation of each cluster.

Let's make a dataframe in which to keep track of the assignments.

```{r}
annotation <- data.frame(
  cluster = unique(seur$seurat_clusters) %>% sort(),
  annotation = NA,
  notes = NA
)
```

For each set of marker genes, we can also perform a number of manual checks to assess cell type annotation.  

First is GO enrichment using `clusterProfiler` package, for which below is a function for quickly looking at top 10 enriched GO categories.
```{r}
# function for GO enrichment
# 1. for a given cluster, prints n.print.rows top enriched GO terms
# 2. Can exclude ribosomal genes (default) because they dominate some GO enrichments
enrichGO2 <- function(clust, ngenes = 100, include.ribo = FALSE, n.print.rows = 10, ...) {
  
  ids.bitr <- bitr(markers[[clust]]$gene[1:ngenes], 
                fromType = "SYMBOL", toType = c("ENTREZID", "SYMBOL"), 
                OrgDb = "org.Hs.eg.db")
    
  if(include.ribo == FALSE)
    features <- ids.bitr$ENTREZID[!(grepl("RP[SL]", ids.bitr$SYMBOL))]
  if(include.ribo == TRUE)
    features <- ids.bitr$ENTREZID

  ego <- enrichGO(
    gene          = features,    
    OrgDb         = org.Hs.eg.db,
    ont           = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff  = 0.01, # default 0.01
    qvalueCutoff  = 0.05, # default 0.05
    readable = TRUE)
  
  dftoprint <- clusterProfiler::simplify(ego)@result[1:n.print.rows, c(2,8)]
  rownames(dftoprint) <- NULL
  
  return(dftoprint)
}
```

We can also check if the cluster is enriched is genes that are labeled as markers genes for a cell type by other studies. [@vento-tormo_single-cell_2018] have provided top 30 marker genes for all cell types identified by them.

```{r message=FALSE}
# read top 30 markers from vento-tormo et al
markers.vento <- readxl::read_excel("../info/from-papers/vento-tormo_2018/41586_2018_698_MOESM1_ESM/Supplementary Table 2.xlsx")

# rename columns
markers.vento <- dplyr::rename(markers.vento,
                               VCT_1 = VCT...2,
                               Tcells_1 = `T cells...5`,
                               Tcells_2 = `T cells...9`,
                               VCT_2 = `VCT...10`,
                               Tcells_3 = `T cells...11`,
                               FB_1 = F1,
                               EVT_1 = `EVT...14`,
                               Tcells_4 = `T cells...15`,
                               EVT_2 = `EVT...16`,
                               NKCD16pos = `Blood NK CD16+`,
                               MO_1 = `MO...22`,
                               MO_2 = `MO...27`,
                               NKCD16neg = `Blood NK CD16-`,
                               FB2 = F2,
                               Endo.m = `Endo (m)`,
                               Endo.L = `Endo L`,
                               Endo.f = `Endo (f)` 
)

# remove ensembl ids
markers.vento <- apply(X = markers.vento, MARGIN = 2, FUN = function(x) gsub("_ENSG\\d+", "", x)) %>% 
  as.data.frame()
```

[@suryawanshi_single-cell_2018] have also provided a list of top30 marker genes in the supplementary data. 
```{r}
markers.surya.vil <- readxl::read_excel("../info/from-papers/suryawanshi_2018/supp-data/aau4788_Data_file_S3.xlsx", sheet = "Villi", range = "A1:G271", trim_ws = TRUE)
markers.surya.dec <- readxl::read_excel("../info/from-papers/suryawanshi_2018/supp-data/aau4788_Data_file_S3.xlsx", sheet = "Deciuda", range = "A1:G331", trim_ws = TRUE)

markers.surya.vil <- dplyr::rename(markers.surya.vil, cluster = `cell type`)

markers.surya.vil$cluster <- paste0("vil.", markers.surya.vil$cluster)
markers.surya.dec$`cluster` <- paste0("dec.", markers.surya.dec$`cluster`)

markers.surya <- rbind(markers.surya.vil, markers.surya.dec)
```


## DSC
Clusters 0 and 18.

### Notes
1. Clusters 0 and 18 highly correlated with each other and most similar to each other than to other clusters. 
2. Both are consistently similar to the decidual stromal cells in both vento and surya datasets. 
3. The marker genes include Prolactin, IGFBP1 etc, which are typical of decidual stromal cells. 
4. These two clusters also arise unabiguously from the decidua samples rather than villi samples.

```{r}
annotation$annotation[annotation$cluster %in% c("clust_00")] <- "dec.DSC_1"
annotation$annotation[annotation$cluster %in% c("clust_18")] <- "dec.DSC_2"

annotation[annotation$cluster %in% paste0("clust_", c("00", "18")), ]
```

### Marker genes plots
```{r fig.asp=1}
top50marker.plots[["clust_00"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_18"]]
```

### Go anrichment
```{r}
enrichGO2("clust_00")
```

### Vento markers
```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$dS3, title = "vento: dS3")
```

### Surya markers
```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "dec.DSC"], title = "surya markers: DSC")
```

## Endothelial cells
Clusters 4 and 21.

### Notes

1. Clusters 4 and 21 are highly correlated. 
2. They consistently get assigned as decidual endothelial cells by both vento and surya datasets.
3. More than 75% of the cells in both of these clusters come from the decidua samples.
4. Markers genes of both clusters are enriched in GO categories related to epithelium development, endothelium development etc. 
5. But these two clusters are not as alike as clusters 0 and 18. They are different in terms of their expression of many endothelium related genes like CD34, PROX1, VWF, SPARCL. Cluster 21 looks more like "typical" endothelial cells. Cluster 4 has low expression of many typical endothelial genes. 
6. Cluster 4 has higher expression of some lymphatic endothelial cells: PROX1, LYVE1. Many genes from the list of surya marker genes for dec.LEC are expressed in cluster 4. This suggests that Cluster 4 may be lymphatic endothelial cells and cluster 21 may be endothelial cells. 

```{r}
annotation$annotation[annotation$cluster %in% c("clust_04")] <- "dec.Endo.L"
annotation$annotation[annotation$cluster %in% c("clust_21")] <- "dec.Endo"

annotation[annotation$cluster %in% paste0("clust_", c("04", "21")), ]
```

### Marker gene plots
```{r, fig.asp=1}
top50marker.plots[["clust_04"]]
```

```{r, fig.asp=1}
top50marker.plots[["clust_21"]]
```

### GO enrichment
```{r}
enrichGO2("clust_04")
```

```{r}
enrichGO2("clust_21")
```

### Vento markers
```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$Endo.m, title = "vento: Endo.m")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$Endo.f, title = "vento: Endo.f")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$Endo.L, title = "vento: Endo.L")
```

### Surya markers
```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "vil.VEC"], title = "surya markers: vil.VEC")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "dec.VEC"], title = "surya markers: dec.VEC")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "dec.LEC"], title = "surya markers: dec.LEC")
```

## EVT
Cluster 13. 

### Notes

1. This cluster is unique in its expression of HLA-G, typical of extravillous trophoblast. 
2. Several other genes often enriched in EVT are also enriched in this cluster: FN1, PAPPA2, DIO2, etc.
3. top 30 markers associated with EVT_2 from vento dataset are also highly and specifically expressed in this cluster. 
4. It is consistently assigned to be EVT based on vento, surya, and pavli datasets. 

Annotation:  
**Cluster 13: Extravillous trophoblast (vil.EVT)**

```{r}
annotation$annotation[annotation$cluster %in% c("clust_13")] <- "vil.EVT"

annotation[annotation$cluster %in% c("clust_13"), ]
```

### Marker genes plot
```{r fig.asp=1}
top50marker.plots[["clust_13"]]
```

### GO enrichment
```{r message=FALSE}
enrichGO2("clust_13")
```

### Vento markers
```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$EVT_1, title = "vento markers: EVT_1")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$EVT_2, title = "vento markers: EVT_2")
```

### Surya markers
```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "vil.EVT"], title = "surya markers: vil.EVT")
```

## VCT and SCT
clusters 16, 17, 23, 24, 25, and 29.

### Notes
For more information on expression patterns and markers of different trophoblast types, see [@liu_single-cell_2018].

1. In addition to cluster 13 (EVT), there are 6 other clusters that are likely trophoblast cell types (they express KRT7 and are related clusters). These 6 clusters are organized into two correlated blocks (see "correlations section above"): (16, 17, 25) and (23, 29, 24) of which the former are likely VCT and the latter SCT according to correlation-based annotation. 
2. Cluster 29 appears to be SCT given its expression of CGA, GDF15, ERVFRD-1 (Syncytin 2) genes. For some reason GO enrichment didn't work for markers of cluster 29, but it does consistently express SCT markers from vento and surya.
3. Cluster 24 doesn't express ERVFRD-1 much, but it does express most other telltale genes of SCT: CSH1, CSH2, CGA, CSHL1, many PSG genes, GH2, PGF, GDF15, etc. And it is not enriched for expression of VCT markers from vento and surya. It's enriched GO terms have to do with hormone production and signaling as expected for SCT. Thus, this is most likely an SCT cluster. 
4. Clusters 16 and 17 don't express many markers of SCT, but do express many vento and surya markers of VCT: PAGE4, PEG10, XAGE3, etc.
5. Cluster 25 has VCT markers as well as a strong signature of cell division genes. This shows up in the GO enrichment as well, where most enriched GO terms are "nuclear division", "chromosome segregation", "cell cycle G1/S phase transition" etc. This is likely a cluster of proliferating VCT cells. Incidentally both [@vento-tormo_single-cell_2018] and [@liu_single-cell_2018] have a separate population of VCT with a cell cycle gene signature (VCT_2 in vento). 
6. Cluster 23 is most similar to SCT of vento but VCT of surya. Indeed this cluster has expression of marker genes for both VCT and SCT (see plots). It expresses PAGE4, PEG10 etc, but also expresses syncytin gene ERVW-1. This could represent a group of cytotrophoblast cells that are in the process of differentiating to sycnytial trophoblast. We will label it as a VCT type. 

```{r}
annotation$annotation[annotation$cluster %in% c("clust_16")] <- "vil.VCT_1"
annotation$annotation[annotation$cluster %in% c("clust_17")] <- "vil.VCT_2"
annotation$annotation[annotation$cluster %in% c("clust_23")] <- "vil.VCT_3"
annotation$annotation[annotation$cluster %in% c("clust_25")] <- "vil.VCT_4"
annotation$annotation[annotation$cluster %in% c("clust_24")] <- "vil.SCT_1"
annotation$annotation[annotation$cluster %in% c("clust_29")] <- "vil.SCT_2"

annotation$notes[annotation$cluster %in% c("clust_25")] <- "proliferating"

annotation[annotation$cluster %in% paste0("clust_", c("16", "17", "23", "25", "24", "29")), ]
```

### Marker gene plots
```{r fig.asp=1}
top50marker.plots[["clust_16"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_17"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_25"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_23"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_24"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_29"]]
```

### GO enrichment
```{r message=FALSE}
enrichGO2("clust_16")
```

```{r message=FALSE}
enrichGO2("clust_17")
```

```{r message=FALSE}
enrichGO2("clust_25")
```

```{r message=FALSE}
enrichGO2("clust_23")
```

```{r message=FALSE}
enrichGO2("clust_24")
```

```{r message=FALSE}
enrichGO2("clust_29")
```

### Vento markers
```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$VCT_1, title = "vento markers: VCT_1")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$VCT_2, title = "vento markers: VCT_2")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$SCT, title = "vento markers: SCT")
```

### Surya markers
```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "vil.VCT"], title = "surya markers: vil.VCT")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "vil.SCT"], title = "surya markers: vil.SCT")
```

### Pavli markers
```{r fig.asp=0.4, warning=FALSE}
DotPlot2(features = c("ADRB1", "PUF60", "SNORDA3A", "PRMT7", "E1F1AY", 
                      "FXYD3", "GRAMD2", "INSL4", "ITGB8", "PAGE4",
                      "SLC13A4", "SLC22A11"), 
         title = "pavli markers: VCT_1")
```


```{r fig.asp=0.6, warning=FALSE}
DotPlot2(features = c("XAGE3", "XAGE2", "SERINC5", "S100P", "RASA1",
                      "MFAP5", "LIN28B", "KRT8", "KRT7", "INS-IGF2", "GCM1",
                      "GATA3", "ERVW-1", "ERVFRD-1", "EGR1", "EFNA1", 
                      "CYP19A1", "PHLDA2", "ACKR2"), 
         title = "pavli markers: VCT_2")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = c("SEMA3B", "CSH2", "CSHL1", "FCGR2A", "GH2", "HBA1", 
                      "HBA2", "HBB", "HBG1", "HBG2", "HLA-DMA", "HLA-DPA1",
                      "HLA-DQB1", "HLA-DRA", "HLA-DRB1", "HPGDS", "CGB8",
                      "LGALS14", "LYVE1", "NPIPB3", "CGB5", "PSG1", "PSG3", 
                      "PSG6", "PSG9", "ZNF117", "ZNF91", "HIST2H2AB",
                      "RAMP2", "MUC20"), 
         title = "pavli markers: SCT")
```

### Liu et al markers for trophoblasts
```{r fig.asp=0.5}
DotPlot2(seur, features = c("GCM1", "CSH1", "KRT7", "TFAP2C", "GATA3", "CDH1", "EGFR", "HLA-G", "MMP2", "RRM2", "CCNB1", "CDK1", "ERVFRD-1", "SLC1A5", "PAGE4", "CGB"), title = "Genes from Liu et al 2018")
```
## Fibroblasts and SMC
Clusters 02, 06, 07, 15, 19, and 33.

### Notes

1. Most of these clusters represent some kind of fibroblast or a closely related cell type. Most clusters are enriched for GO terms related to ECM regulation. 
2. Clusters 02 and 15 marker genes are also highly expressed in clusters 00 and 18. Cluster 15 has more than 75% cells coming from decidua samples, and cluster 02 has over 50% cells coming from decidua samples. Decidual stromal cells are known to be actually a few different populations. Even in Vento-tormo et al, they are labeled as dS1--3, of which dS3 are the canonical DSC with PRL and IGFBP1, while dS1 and dS2 are closely related to DSC and are derived from endometrial stromal fibroblasts. These things together suggest that clusters 02 and 15 are the non-PRL varieties of DSC. To distinguish them from DSC, we can call them decidual fibroblasts. 
3. Marker genes of 06, 07, and 33 show highly correlated gene expressin, i.e. the markers on each of these clusters are also highly expressed in the other two clusters, suggesting that these three clusters a closely related cell populations. All three of these clusters are largely (over 75% of the cells) derived from villi samples, and they are fibroblast-like cells. Thus, these are likely villous fibroblasts.
4. Cluster 19 is most similar to PV (perivascular) clusters from vento and SMC (smooth muscle cell) cluster from surya. Indeed, its markers include smooth muscle genes like MGP, MYH11, MYL9 etc, and enriched GO terms for this cluster include "muscle contraction" and "artery morphogenesis". Thus, this cluster is most likely of perivascular smooth muscle cells of decidual origin (over 75% cells from decidua samples). 

```{r}
annotation$annotation[annotation$cluster %in% c("clust_02")] <- "dec.FB_1"
annotation$annotation[annotation$cluster %in% c("clust_15")] <- "dec.FB_2"
annotation$annotation[annotation$cluster %in% c("clust_06")] <- "vil.FB_1"
annotation$annotation[annotation$cluster %in% c("clust_07")] <- "vil.FB_2"
annotation$annotation[annotation$cluster %in% c("clust_33")] <- "vil.FB_3"
annotation$annotation[annotation$cluster %in% c("clust_19")] <- "dec.SMC"

annotation[annotation$cluster %in% paste0("clust_", c("02", "15", "06", "07", "33", "19")), ]
```

### Marker gene plots
```{r fig.asp=1}
top50marker.plots[["clust_02"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_15"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_06"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_07"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_19"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_33"]]
```

### GO enrichment
```{r message=FALSE}
enrichGO2("clust_02")
```

```{r message=FALSE}
enrichGO2("clust_15")
```

```{r message=FALSE}
enrichGO2("clust_06")
```

```{r message=FALSE}
enrichGO2("clust_07")
```

```{r message=FALSE}
enrichGO2("clust_19")
```

```{r message=FALSE}
enrichGO2("clust_33")
```

### Vento markers

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$PV1, title = "vento markers: PV1")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$PV2, title = "vento markers: PV2")
```

### Surya markers
```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "vil.FB1"], title = "surya markers: vil.FB1")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "dec.SMC"], title = "surya markers: dec.SMC")
```

## B cells
Cluster 20.

### Notes
This is most likely a cluster of B cells. Most markers genes are related to B cells: many immunoglobulin gnees, SPIB, MS4A1, CD79A, CD79B, BANK1, etc. It also expresses many genes that are markers for Plasma cluster from vento dataset. GO enrichment also agrees.

```{r}
annotation$annotation[annotation$cluster %in% c("clust_20")] <- "Bcell"

annotation[annotation$cluster %in% paste0("clust_", c("20")), ]
```

### Marker genes plot
```{r fig.asp=1}
top50marker.plots[["clust_20"]]
```

### GO enrichment
```{r message=FALSE}
enrichGO2("clust_20")
```

### Vento markers
```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$Plasma, title = "vento markers: Plasma")
```

## Other lymphocytes
Clusters  05, 09, 10, 14, 30, and 32.

### Notes
For clues on marker genes for various types of lymphocytes see [@pizzolato_single-cell_2019].

1. These are likely non-B lymphoid cells. 
2. Clusters 09 and 32 are most similar to Tcells from both vento and surya. Given the expression of CD3 genes, that appears to be the correct assignment. 
     + The clusters also express TRAC and TRBC1 (constant regions of TCR alpha and beta), making them alpha-beta T cells. 
     + It's hard to tell if they are CD4 or CD8---both CD4 and CD8 genes are expressed by a small fraction of cells in both clusters, but CD8 genes expressed at a higher level. 
     + GO enrichment for both include terms related to T cell differentiation, T cell activation etc. 
     + These clusters contain cells from both decidua and villi samples. 
3. Cluster 05 is most similar to NK cells from vento and surya. But based on the lack of NKG7 and NCAM1 expression and expression of CD3 genes, TRAC, and TRBC1 this is likely a T cell cluster. One of its marker genes being GZMA and its overall similarity with NK cells suggest that this may be a cytotoxic T cell. Consistently, among its enriched GO terms are "T cell mediated cytotoxicity" and "Cell killing". This cluster mostly contains cells from decidua samples.
4. Clusters 10, 14, and 30 are NK cells based on the expression of NKG7 and NCAM1. 
     + These clusters are also express GZMA, GZMB, GNLY, PRF1, KLRB1, KLRC1, KLRD1. 
     + Cluster 30 is enriched for cell cycle genes, so represents proliferating cells. Vento and surya datasets also have a cluster of proliferating NK cells (dNK.p in vento and NK2 in surya). +
     + Cluster 30 in unambiguously of decidual sample origin. Clusters 10 and 14 are mixed. 
     
```{r}
annotation$annotation[annotation$cluster %in% c("clust_05")] <- "Tcell_1"
annotation$annotation[annotation$cluster %in% c("clust_09")] <- "Tcell_2"
annotation$annotation[annotation$cluster %in% c("clust_10")] <- "NK_1"
annotation$annotation[annotation$cluster %in% c("clust_14")] <- "NK_2"
annotation$annotation[annotation$cluster %in% c("clust_30")] <- "NK_3"
annotation$annotation[annotation$cluster %in% c("clust_32")] <- "Tcell_3"

annotation$notes[annotation$cluster %in% c("clust_05")] <- "cytotoxic"
annotation$notes[annotation$cluster %in% c("clust_30")] <- "proliferating"


annotation[annotation$cluster %in% paste0("clust_", c("05", "09", "10", "14", "30", "32")), ]
```

### Lymphocyte candidate genes
```{r fig.asp=0.5}
DotPlot2(features = c("NCAM1", "NKG7", "CD3D", "CD3G", "CD3E", "TRAC", "TRBC1", "TRGC2", "TRDC", "CD8A", "CD8B", "CD4"), title = "Lymphocyte candidate genes")
```

### Marker gene plots
```{r fig.asp=1}
top50marker.plots[["clust_05"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_09"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_10"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_14"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_30"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_32"]]
```

### GO enrichment
```{r message=FALSE}
enrichGO2("clust_05")
```

```{r message=FALSE}
enrichGO2("clust_09")
```

```{r message=FALSE}
enrichGO2("clust_10")
```

```{r message=FALSE}
enrichGO2("clust_14")
```

```{r message=FALSE}
enrichGO2("clust_30")
```

```{r message=FALSE}
enrichGO2("clust_32")
```

### Vento markers

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$Tcells_2, title = "vento markers: Tcells_2")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$Tcells_4, title = "vento markers: Tcells_4")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$dNK1, title = "vento markers: dNK1")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$dNKp, title = "vento markers: dNKp")
```


### Surya markers
```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "dec.NK1"], title = "surya markers: dec.NK1")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "dec.NK2"], title = "surya markers: dec.NK2")
```

## Myeloid cells
Clusters 01, 03, 08, 11, 12, 22, 26, 27, 28, 31, and 34.

### Notes

1. Clusters 01, 22, 31, 34, and 27 form a group in the hierarchical clustering above, and they all have most cells (more than 75%) coming from villi samples. 
    + All five clusters are highly highly correlated with each other, especially clusters 01, 31, and 34. 
    + All clusters express various macrophage genes: CD68, CD163, MRC1, MAF, C1QC, C1QA, FOLR2 etc; and more so clusters 01, 31, and 34. 
    + Clusters 01 and 31 are most likely villous macrophages (i.e. Hofbauer cells), but the other cluster need more investigation. 
    + Clusters 22 and 27, even though correlated with 01 and 34, have marker genes that are not that highly expressed in themselves but are in 01 and 34. Looking at the UMAP plot below, it's clear that both of these clusters are not as homogenous as the other. The cells are spread between the big myeloid blob and the big fibroblast blob. 
    + However, these don't appear to be clusters of different types of cells. That is, fibroblast genes (COL3A1, COL3A3) and macrophage genes are coexpressed in the same cells in cluster 27 (see plot below). And macrophage genes are broadly expressed in cluster 22, even in cells that seem to be embedded with fibroblast cells. Clusters 22 and 27 might represent cell types that are macrophages but with certain fibroblast-like properties. 
    + Looking at cluster 34 markers genes, it interestingly expresses many erythrocyte lineage genes (HBA1, HBA2, HBM, CD1, HBG1, CD235a/GYPA) but it also expresses macrophage genes like CSF1R, FOLR2, MRC1, CD163 etc. If we plot the expression of these genes on UMAP (plotting only clust_34, see below), we see that majority of the cells express macrophage genes, while the erythrocyte genes are expressed in a small number of cells (except HBG2), suggesting that this cluster is made of two cell types. Since majority of the cells in this cluster are macrophage cells, we will label it as such, but should the need arise in further analysis, we can subcluster it further to separate the two cell types out.
2. Cluster 26 is most likely of erythrocytes (or some cell type from the erythrocyte lineage). 
    + The surya dataset does contain what they call an erythroblast cluster, but in erythropoeisis is expected in the placenta in the first trimester. In our term samples, I am not sure if erythrocyte progenitors are expected to be present. 
    + Marker genes, similarity with surya EB cluster, and GO enrichement (which include "erythrocyte development", "oxygen transport", etc) are all consistent with each other. 
    + Most of the cells in this cluster come from villi samples, so fetal origin.
3. Cluster 28 most likely is made of typical granulocytes. 
    + It expresses the granulocyte marker gene CEACAM8, in addition to CEACAM6, ELANE, and MPO. 
    + GO enrichment is consistent.
    + All marker genes are consistent: LYZ, DEFA3, DEFA4, S100A8, NCF1, etc. 
4. Cluster 11 and 12 are most likely APC.
    + Of maternal origin. Cluster 12 mostly comes from decidua sample, cluster 11 has mixed contribution, in terms of its gene expression it's very similar to cluster 12. 
    + Very similar to each other. Among vento cluster, most similarity to macrophage clusters. Among surya clusters, most similarity to maternal DC and maternal APC clusters.
    + Expression of classical macrophage/DC markers like CD68, CD163. In addition most of the marker genes of this cluster have to do with antigen presentation. Many of them are MHC-II genes. 
    + Not sure if they are macrophages or DCs, but to avoid committing to one conclusion over the other, we can just label them as APCs. 
5. Clusters 3 and 8 are difficult.
    + Low, if any, expression of CD68 and CD163, thus not quite macrophages. But at least cluster 8 is quite similar to macrophages in overall gene expression (see hierarchical clustering above).
    + Low, if any, expression of HLA genes, thus not quite APCs.
    + No expression of typical granulocyte markers: CEACAM8, CEACAM6, MPO, ELANE. But many markers of these clusters sound "granulocytic". This is evident in GO enrichment as well. In addition, cluster 3 is most similar to cluster 28 (see hierarchical clustering above) which is clearly a granulocyte cluster.
    + Cluster 08 is most similar to the MO_1 and MO_2 clusters from vento. Cluster 3 had low similarity with any of the clusters from vento and surya. 
    + Both clusters express MNDA, which is typically only expressed in cells in the monocyte-granulocyte lineage. 
    + Both clusters express LYZ, S100A9, S100A8, which I think are also monocyte genes.
    + These might be monocytes but they don't express CD14. 
    + It's in fact possible to have CD14--ve monocyte, it turns out [@mukherjee_non-classical_2015]. See introduction of [@sampath_monocyte-subsets_2018], which summarizes observations from [@villani_single-cell_2017]. These papers point to non-CD14 monocyte populations. In fact most genes in the Mono_3 cluster of [@villani_single-cell_2017] are expressed in cluster 03, e.g. MXD1, VNN2, CXCR2.
    + Overall it looks like clusters 3 and 8 are two types of monocytes. (at the very least they are from the monocyte-granulocyte lineage.) 
    
```{r}
annotation$annotation[annotation$cluster %in% c("clust_01")] <- "vil.Hofb_1"
annotation$annotation[annotation$cluster %in% c("clust_22")] <- "vil.Hofb_2"
annotation$annotation[annotation$cluster %in% c("clust_27")] <- "vil.Hofb_3"
annotation$annotation[annotation$cluster %in% c("clust_31")] <- "vil.Hofb_4"
annotation$annotation[annotation$cluster %in% c("clust_34")] <- "vil.Hofb_5"

annotation$annotation[annotation$cluster %in% c("clust_03")] <- "Mono_1"
annotation$annotation[annotation$cluster %in% c("clust_08")] <- "Mono_2"
annotation$annotation[annotation$cluster %in% c("clust_11")] <- "APC_1"
annotation$annotation[annotation$cluster %in% c("clust_12")] <- "APC_2"
annotation$annotation[annotation$cluster %in% c("clust_26")] <- "vil.Ery"
annotation$annotation[annotation$cluster %in% c("clust_28")] <- "Gran"

annotation$notes[annotation$cluster %in% c("clust_27")] <- "also fibroblast signature"
annotation$notes[annotation$cluster %in% c("clust_34")] <- "likely also contains erythro"
annotation$notes[annotation$cluster %in% c("clust_03")] <- "unsure"
annotation$notes[annotation$cluster %in% c("clust_08")] <- "unsure"

annotation[annotation$cluster %in% paste0("clust_", 
                                          c("01", "03", "08", "11", "12",
                                            "22", "26", "27", "28", "31", "34")), ]
```

### Myeloid candidate genes
```{r fig.asp=0.5}
DotPlot2(features = c("PTPRC", # pan
                      "CD33", "CD14", "CD68", "CD163", "CLEC10A", "HLA-DRA", # apc
                      "CEACAM8", "CEACAM6", "MPO", "ELANE", # granulocytes
                      "FCGR3A", # inflammatory monocytes?
                      "GYPA" # erythro
                      ), 
         title = "Myeloid candidate genes")
```

### Hofb clusters on UMAP
```{r fig.width=6}
DimPlot(seur, 
        cols = c(rep("Grey90", 30), "#aec7e8", "#ff9e4a", "#2ca02c", "#ed665d", "#9467bd"), 
        order = (rev(c("clust_01", "clust_22", "clust_27", "clust_31", "clust_34")))) + 
  theme_few() +
  coord_fixed()
```
### clust_22 markers on UMAP
```{r fig.asp=0.5}
DefaultAssay(seur) <- "SCT"
p <- FeaturePlot(seur, features = c("CD163", "CSF1R", "C1QA", "MRC1", "CD14", "FOLR2"), 
                 cells = rownames(seur@meta.data)[seur@meta.data$seurat_clusters == "clust_22"],
                 coord.fixed = TRUE, ncol = 3, combine = FALSE) 

p[[1]] + p[[2]] + p[[3]] + p[[4]] + p[[5]] + p[[6]] + 
  plot_layout(ncol = 3) +
  plot_annotation(title = "clust_22") &
  theme_bw() +
  theme(panel.grid.major = element_line(size = 0.25),
        panel.grid.minor = element_blank(),
        legend.key.size = unit(0.75, "line"),
        axis.title = element_text(size = 6),
        plot.title = element_text(size = 10)) 
```

### clust_27 markers on UMAP
```{r}
DefaultAssay(seur) <- "SCT"
p <- FeaturePlot(seur, features = c("COL3A1", "CD14", "MRC1", "LYVE1", "COL1A1"), cells = rownames(seur@meta.data)[seur@meta.data$seurat_clusters == "clust_27"],
                 coord.fixed = TRUE, ncol = 3) 
p[[1]] + p[[2]] + p[[3]] + p[[4]] + p[[5]] + 
  plot_layout(ncol = 3) +
  plot_annotation(title = "clust_27") &
  theme_bw() +
  theme(panel.grid.major = element_line(size = 0.25),
        panel.grid.minor = element_blank(),
        legend.key.size = unit(0.75, "line"),
        axis.title = element_text(size = 6),
        plot.title = element_text(size = 10)) 
```

### clust_34 markers on UMAP
```{r}
DefaultAssay(seur) <- "SCT"
p <-  FeaturePlot(seur, 
                  features = c("GYPA", "HBM", "HBG2", "HBG1", "ALAS2", "AHSP", 
                               "MRC1", "CD163", "CSF1R"), 
                  cells = rownames(seur@meta.data)[seur@meta.data$seurat_clusters == "clust_34"],
                  coord.fixed = FALSE, ncol = 3)
p[[1]] + p[[2]] + p[[3]] + p[[4]] + p[[5]] + p [[6]] + p[[7]] + p[[8]] + p[[9]] + 
plot_layout(ncol = 3) +
  plot_annotation(title = "clust_34") &
  theme_bw() +
  theme(panel.grid.major = element_line(size = 0.25),
        panel.grid.minor = element_blank(),
        legend.key.size = unit(0.75, "line"),
        axis.title = element_text(size = 6),
        plot.title = element_text(size = 10)) 
```

### Marker gene plots
```{r fig.asp=1}
top50marker.plots[["clust_01"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_03"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_08"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_11"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_12"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_22"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_26"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_27"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_28"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_31"]]
```

```{r fig.asp=1}
top50marker.plots[["clust_34"]]
```

### GO enrichment
```{r message=FALSE}
enrichGO2("clust_01")
```

```{r message=FALSE}
enrichGO2("clust_03")
```

```{r message=FALSE}
enrichGO2("clust_08")
```

```{r message=FALSE}
enrichGO2("clust_11")
```

```{r message=FALSE}
enrichGO2("clust_12")
```

```{r message=FALSE}
enrichGO2("clust_22")
```

```{r message=FALSE}
enrichGO2("clust_26")
```

```{r message=FALSE}
enrichGO2("clust_27")
```

```{r message=FALSE}
enrichGO2("clust_28")
```

```{r message=FALSE}
enrichGO2("clust_31")
```

```{r message=FALSE}
enrichGO2("clust_34")
```

### Vento markers
```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$dM1, title = "vento markers: dM1")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$dM2, title = "vento markers: dM2")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$HB, title = "vento markers: HB")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$M3, title = "vento markers: M3")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$MO_1, title = "vento markers: MO_1")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$MO_2, title = "vento markers: MO_2")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$DC1, title = "vento markers: DC1")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.vento$DC2, title = "vento markers: DC2")
```

### Surya markers
```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "vil.HC"], title = "surya markers: vil.HC")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "vil.EB"], title = "surya markers: vil.EB")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "dec.APC"], title = "surya markers: dec.APC")
```

```{r fig.asp=0.8, warning=FALSE}
DotPlot2(features = markers.surya$gene[markers.surya$cluster == "vil.HC"], title = "surya markers: vil.HC")
```

# Final annotations
## Merge similar clusters
In addition to the annotations so far, we can have a second set of coarser-grained annotations so that we have fewer clusters to deal with. For this we will merge some similar clusters---mostly non-immune clusters, leaving all immune clusters intact. 

```{r}
annotation$annotation_merged <- annotation$annotation

annotation$annotation_merged[annotation$annotation %in% paste0("vil.Hofb_", c(1:5))] <- "vil.Hofb"
annotation$annotation_merged[annotation$annotation %in% paste0("dec.DSC_", c(1:2))] <- "dec.DSC"
annotation$annotation_merged[annotation$annotation %in% paste0("dec.FB_", c(1:3))] <- "dec.FB"
annotation$annotation_merged[annotation$annotation %in% paste0("vil.FB_", c(1:3))] <- "vil.FB"
annotation$annotation_merged[annotation$annotation %in% paste0("APC_", c(1:2))] <- "APC"
annotation$annotation_merged[annotation$annotation %in% paste0("vil.VCT_", c(1:4))] <- "vil.VCT"
annotation$annotation_merged[annotation$annotation %in% paste0("vil.SCT_", c(1:2))] <- "vil.SCT"
annotation$annotation_merged[annotation$annotation %in% c("dec.Endo", "dec.Endo.L")] <- "dec.Endo"
```

```{r}
# rearrange columns and write annotations to csv
annotation <- annotation[, c("cluster", "annotation", "annotation_merged", "notes")]

write.csv(annotation, "../results/02_annotation/files/annotations.csv", row.names = FALSE)
```


```{r}
annotation %>% kable(caption = "Final annotations of all clusters") %>% kable_styling(full_width = FALSE)
```
## Add annotations to `Seurat` object
```{r}
# add annotations
seur@meta.data$annotation <- annotation$annotation[match(
  x = seur@meta.data$seurat_clusters,
  table = annotation$cluster)
  ]

seur@meta.data$annotation_merged <- annotation$annotation_merged[match(
  x = seur@meta.data$seurat_clusters, 
  table = annotation$cluster)
  ]

```


```{r}
# write seurat object
saveRDS(seur, file = "../data/seurat-object_annotated.rds")
```

## Plots
### UMAP: all annotations
```{r, fig.asp=1, fig.width=7}
Idents(seur) <- seur@meta.data$annotation
DimPlot(seur, label = TRUE, repel = TRUE, label.size = 4, shuffle = TRUE) +
  coord_fixed() +
  theme_minimal() +
  theme(legend.position = "none",
        panel.grid = element_blank(),
        axis.title = element_text(size = 8))

ggsave(last_plot(), filename = "../results/02_annotation/plots/umap_annotated.pdf", 
       device = "pdf", width = 7, height = 7, units = "in")
```
### UMAP: merged annotations
```{r, fig.asp=1, fig.width=7}
Idents(seur) <- seur@meta.data$annotation_merged
DimPlot(seur, label = TRUE, repel = TRUE, label.size = 4, shuffle = TRUE) +
  coord_fixed() +
  theme_minimal() +
  theme(legend.position = "none",
        panel.grid = element_blank(),
        axis.title = element_text(size = 8))

ggsave(last_plot(), filename = "../results/02_annotation/plots/umap_annotated_merged.pdf", 
       device = "pdf", width = 7, height = 7, units = "in")
```

# Session Info
```{r}
sessionInfo()
```

# References

