---
title: "Receptor-ligand interactions using CellPhoneDB"
author: "Arun Chavan"
output:
  html_notebook:
    toc: yes
    toc_float: yes
    number_sections: true
bibliography: ../refs.bib
---
Started: 2020-10-20   
Last edited: `r format(Sys.time())`

```{r message=FALSE}
library(tidyverse)
```

# Overview
The goal of this analysis is to find out if cellular communication mediated by receptor-ligand interactions is different between control and covid samples. For this we use [CellPhoneDB](https://www.cellphonedb.org/) tool developed by [@efremova_cellphonedb_2020]. CellPhoneDB is both a repository a curated receptors, ligands and their interactions, and a python package for extracting enriched receptor-ligand interactions between pairs of clusters in single cell RNA-seq data. 

# Running CellPhoneDB
## How it works
For every receptor-ligand pair and for every pair of clusters in the data, CellPhoneDB calculates the mean expression of the receptor and the ligand between the two clusters in question (the mean is set to 0 if either receptor or ligand expression is 0). To test whether the interaction is specific to these two clusters, it shuffles the cluster labels of cells a defined number of times, each time calculating the mean expression of R and L to generate a null distribution. Based on this null distribution, it calculates a *p* value for observed mean expression. Note that a non-significant *p* value does not mean that the interaction is not present, but that it is not highly specific to the pair of clusters.

## Running it
I ran it using the python package, separately for covid and control cells from our dataset, following the instruction from here: https://github.com/Teichlab/cellphonedb. The code used for generating the input files and for running CellPhoneDB is in `./code/`. Generation of input files had to be done on the cluster because my computer could not handle it. Running CellPhoneDB also should have ideally been done on the cluster but because I did not have the R environment properly set up on the cluster, the final steps (plot generation using `pheatmap`) failed, so I ended up running it on my computer. The data needed to be subsampled (subsampled both control and covid to 5000 cells) for it to run in a reasonable time---it took ~3 hours per sample after subsampling. The output is saved in `./results/06_cellphonedb/out`. 

## Output files
```{r}
cntrl <- list()
cntrl$count <- read.table("../results/06_cellphonedb/out/cntrl/count_network.txt", header = TRUE, sep = "\t", quote = "")
cntrl$means <- read.table("../results/06_cellphonedb/out/cntrl/means.txt", header = TRUE, sep = "\t", quote = "")
cntrl$pvals <- read.table("../results/06_cellphonedb/out/cntrl/pvalues.txt", header = TRUE, sep = "\t", quote = "")

covid <- list()
covid$count <- read.table("../results/06_cellphonedb/out/covid/count_network.txt", header = TRUE, sep = "\t", quote = "")
covid$means <- read.table("../results/06_cellphonedb/out/covid/means.txt", header = TRUE, sep = "\t", quote = "")
covid$pvals <- read.table("../results/06_cellphonedb/out/covid/pvalues.txt", header = TRUE, sep = "\t", quote = "")
```

# Plots
## Number of interactions
We'll remake the plots using CellPhoneDB output files. The first plot we'll make is of the number of interactions between each pair of cell types. 

### functions
```{r}
reorder_clusters <- function(p.dat) {
  
  # clustering for ordering
  cdat <- pivot_wider(data =  p.dat, id_cols = "SOURCE", names_from = "TARGET", values_from = "count")
  
  cdat <- as.data.frame(cdat)
  rownames(cdat) <- cdat$SOURCE
  cdat$SOURCE <- NULL
  
  cor.matrix <- cor(t(cdat))
  
  # reorder correlation matrix based on clustering
  dd <- as.dist((1 - cor.matrix)) 
  hc <- hclust(dd, method = 'complete')
  cdat  <- cdat[hc$order, ]
  
  p.dat$SOURCE <- factor(p.dat$SOURCE, levels = rownames(cdat))
  p.dat$TARGET <- factor(p.dat$TARGET, levels = rownames(cdat))
  
  return(p.dat)
}
```

```{r}
theme_heatmap <- theme_bw() +
  theme(axis.title = element_blank(),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.key.size = unit(0.75, "lines"))
```


### Control
```{r}
cntrl$count %>% 
  filter(!(SOURCE %in% c("dec.Gran", "dec.Bcells", "vil.Ery", "dec.Tcell_3"))) %>% 
  filter(!(TARGET %in% c("dec.Gran", "dec.Bcells", "vil.Ery", "dec.Tcell_3"))) %>% 
  reorder_clusters() %>% 
ggplot(., aes(SOURCE, TARGET)) +
  geom_tile(aes(fill = count)) +
  scale_fill_gradient(low = "white", high = "#d6604d") +
  coord_fixed() +
  theme_heatmap
```

### Covid
```{r}
covid$count %>% 
  filter(!(SOURCE %in% c("dec.Gran", "dec.Bcells", "vil.Ery", "dec.Tcell_3"))) %>% 
  filter(!(TARGET %in% c("dec.Gran", "dec.Bcells", "vil.Ery", "dec.Tcell_3"))) %>% 
  reorder_clusters() %>% 
  ggplot(., aes(SOURCE, TARGET)) +
  geom_tile(aes(fill = count)) +
  scale_fill_gradient(low = "white", high = "#d6604d") +
  coord_fixed() +
  theme_heatmap
```

### Difference between covid and control
Now we'll plot the difference in the number of interactions between covid and control cells. 

1. The plots below show that in covid samples the number of interactions between immune cells increase. 
2. This can be seen in both plots, absolute number of increased interactions as well as fold change.

```{r}
# create a dataset of differences
count.diff <- inner_join(x = cntrl$count, y = covid$count, 
                         by = c("SOURCE", "TARGET"),
                         suffix = c(".cntrl", ".covid"))
count.diff$diff <- count.diff$count.covid - count.diff$count.cntrl
count.diff$fc <- count.diff$count.covid / count.diff$count.cntrl
```

```{r}
# plot difference in number of interactions (covid - control)
p.dat <- count.diff %>% 
  filter(!(SOURCE %in% c("dec.Gran", "dec.Bcells", "vil.Ery", "dec.Tcell_3"))) %>% 
  filter(!(TARGET %in% c("dec.Gran", "dec.Bcells", "vil.Ery", "dec.Tcell_3"))) %>%
  select(SOURCE, TARGET, diff) %>% 
  arrange(diff)

p.dat$SOURCE <- factor(p.dat$SOURCE, levels = rev(p.dat$SOURCE) %>% unique())
p.dat$TARGET <- factor(p.dat$TARGET, levels = rev(p.dat$SOURCE) %>% unique())
  
ggplot(p.dat, aes(SOURCE, TARGET)) +
  geom_tile(aes(fill = diff)) +
  scale_fill_gradient2(low = "#4393c3", mid = "white", high = "#d6604d", 
                       midpoint = 0,
                       name = "diff = covid - control") +
  coord_fixed() +
  theme_heatmap
```


```{r}
# plot log fold change in number of interactions
p.dat <- count.diff %>% 
  filter(!(SOURCE %in% c("dec.Gran", "dec.Bcells", "vil.Ery", "dec.Tcell_3"))) %>% 
  filter(!(TARGET %in% c("dec.Gran", "dec.Bcells", "vil.Ery", "dec.Tcell_3"))) %>%
  select(SOURCE, TARGET, fc) %>%
  mutate(logfc = log(fc)) %>% 
  arrange(logfc)

p.dat$SOURCE <- factor(p.dat$SOURCE, levels = rev(p.dat$SOURCE) %>% unique())
p.dat$TARGET <- factor(p.dat$TARGET, levels = rev(p.dat$SOURCE) %>% unique())
  
ggplot(p.dat, aes(SOURCE, TARGET)) +
  geom_tile(aes(fill = logfc)) +
  scale_fill_gradient2(low = "#4393c3", mid = "white", high = "#d6604d", 
                       midpoint = 0,
                       name = "log(covid/control)") +
  coord_fixed() +
  theme_heatmap
```

## Dotplots of mean expression and p values
CellPhoneDB output contains a dot plot where gene pairs are on one axis, cell type pairs are on another axis, and dots are coloured by mean expression and sized by pvalue. The standard plot from the output is huge for our data because there are 21*21 cell type pairs. But this plot is nice for finding out exactly which ligand-receptor interactions are enriched between which cell type pairs, which the above heatmaps don't tell us. We need a concise version of the dotplot, which we can arrive at as follows:

1. Given than interactions among immune cell types increase in covid, we can try to plot only the cell type pairs with these immune cells. 
2. Instead of plotting them separately for control and covid, we can plot the ligand receptor interactions that are present in covid samples but absent in control samples. 

### Prepare data

```{r}
# melt means and pvals for cntrl and covid
cntrl$means.long <- pivot_longer(data = cntrl$means, 
                                 cols = 12:ncol(cntrl$means), 
                                 names_to = "celltype.pair", 
                                 values_to = "mean")
cntrl$pvals.long <- pivot_longer(data = cntrl$pvals, 
                                 cols = 12:ncol(cntrl$pvals), 
                                 names_to = "celltype.pair", 
                                 values_to = "pval")
cntrl$means.pval <- inner_join(cntrl$means.long, cntrl$pvals.long)

covid$means.long <- pivot_longer(data = covid$means, 
                                 cols = 12:ncol(covid$means), 
                                 names_to = "celltype.pair", 
                                 values_to = "mean")
covid$pvals.long <- pivot_longer(data = covid$pvals, 
                                 cols = 12:ncol(covid$pvals), 
                                 names_to = "celltype.pair", 
                                 values_to = "pval")
covid$means.pval <- inner_join(covid$means.long, covid$pvals.long)
```

```{r}
# subset to exclude non-significant interactions
cntrl$means.pval <- filter(cntrl$means.pval, pval < 0.05)
covid$means.pval <- filter(covid$means.pval, pval < 0.05)
```

```{r}
# interactions specific to covid
covid$means.pval <- covid$means.pval %>%  
  mutate("interaction_celltype.pair" = paste0(id_cp_interaction, "_", celltype.pair)) 
cntrl$means.pval <- cntrl$means.pval %>% 
  mutate("interaction_celltype.pair" = paste0(id_cp_interaction, "_", celltype.pair))

covid$unique <- covid$means.pval %>% 
  filter(!(interaction_celltype.pair %in% cntrl$means.pval$interaction_celltype.pair))
```

```{r}
# filter to keep only immune cells
celltypes.keep <- c("dec.NK_1", "dec.NK_2", "dec.Tcell_1", "dec.Tcell_2", "dec.APC", "dec.Mono_1", "dec.Mono_2")

pairs.keep <- c(
  paste0(celltypes.keep[1], ".", celltypes.keep),
  paste0(celltypes.keep[2], ".", celltypes.keep),
  paste0(celltypes.keep[3], ".", celltypes.keep),
  paste0(celltypes.keep[4], ".", celltypes.keep),
  paste0(celltypes.keep[5], ".", celltypes.keep),
  paste0(celltypes.keep[6], ".", celltypes.keep),
  paste0(celltypes.keep[7], ".", celltypes.keep)
)

covid$unique.immune <- covid$unique %>% 
  filter(celltype.pair %in% pairs.keep)
```

```{r}
# subset further to keep only those celltype pairs among immune cell types for which number of interactions in covid are more than in control.
count.diff$celltype.pair <- paste0(count.diff$SOURCE, ".", count.diff$TARGET)

covid$unique.immune <- covid$unique.immune %>% 
  filter(celltype.pair %in% count.diff$celltype.pair[count.diff$diff > 0])
```


### Plot

```{r fig.asp=1.9, fig.width=5}
ggplot(covid$unique.immune, aes(celltype.pair, interacting_pair)) +
  geom_point(aes(color = log(mean), size = -log10(pval+0.001)), shape = 19) +
  scale_color_gradient(low = "#fdd0a2", high = "#d94801") +
  scale_size(range = c(1, 4)) +
  theme_heatmap +
  theme(panel.grid = element_line(size = 0.35))
```



# Session Info
```{r}
sessionInfo()
```

# References
